posthog/hogql_queries/insights/trends/calendar_heatmap_query_runner.py

```diff
@@ -0,0 +1,334 @@
+from posthog.hogql_queries.query_runner import QueryRunner
+from posthog.hogql_queries.utils.query_date_range import QueryDateRange
+from posthog.models import Team
+from posthog.models.action.action import Action
+from posthog.models.filters.mixins.utils import cached_property
+from posthog.models.property_definition import PropertyDefinition
+from posthog.schema import (
+    ActionConversionGoal,
+    ActionsNode,
+    CachedCalendarHeatmapQueryResponse,
+    CustomEventConversionGoal,
+    DashboardFilter,
+    DataWarehouseNode,
+    EventsNode,
+    HogQLQueryModifiers,
+    IntervalType,
+    CalendarHeatmapQuery,
+    CalendarHeatmapResponse,
+    EventsHeatMapDataResult,
+    EventsHeatMapRowAggregationResult,
+    EventsHeatMapColumnAggregationResult,
+    EventsHeatMapStructuredResult,
+)
+
+SEPARATOR = "','"
+
+templateUniqueUsers = """
+SELECT
+    mapKeys(query.hoursAndDays) as hoursAndDaysKeys,
+    mapValues(query.hoursAndDays) as hoursAndDaysValues,
+    mapKeys(query.hours) as hoursKeys,
+    mapValues(query.hours) as hoursValues,
+    mapKeys(query.days) as daysKeys,
+    mapValues(query.days) as daysValues,
+    query.total
+FROM (
+    SELECT
+        uniqMap(map(concatWithSeparator({separator},toString(toDayOfWeek(uniqueSessionEvents.timestamp)),toString(toHour(uniqueSessionEvents.timestamp))), uniqueSessionEvents.person_id)) as hoursAndDays,
+        uniqMap(map(toHour(uniqueSessionEvents.timestamp), uniqueSessionEvents.person_id)) as hours,
+        uniqMap(map(toDayOfWeek(uniqueSessionEvents.timestamp), uniqueSessionEvents.person_id)) as days,
+        uniq(person_id) as total
+    FROM (
+        SELECT
+            any(events.person_id) as person_id,
+            session.session_id as session_id,
+            min(session.$start_timestamp) as timestamp
+        FROM events
+        WHERE and(
+            {event_expr},
+            {all_properties},
+            {test_account_filters}
+        )
+        GROUP BY session_id
+    ) as uniqueSessionEvents
+    WHERE {current_period}
+) as query
+"""
+
+templateAllUsers = """
+SELECT
+    mapKeys(query.hoursAndDays) as hoursAndDaysKeys,
+    mapValues(query.hoursAndDays) as hoursAndDaysValues,
+    mapKeys(query.hours) as hoursKeys,
+    mapValues(query.hours) as hoursValues,
+    mapKeys(query.days) as daysKeys,
+    mapValues(query.days) as daysValues,
+    query.total
+FROM (
+    SELECT
+        sumMap(map(concatWithSeparator({separator},toString(toDayOfWeek(events.timestamp)) ,toString(toHour(events.timestamp))), 1)) as hoursAndDays,
+        sumMap(map(toHour(events.timestamp), 1)) as hours,
+        sumMap(map(toDayOfWeek(events.timestamp), 1)) as days,
+        count(*) as total
+    FROM events
+    WHERE and(
+        {event_expr},
+        {all_properties},
+        {test_account_filters},
+        {current_period}
+    )
+) as query
+"""
+
+
+class CalendarHeatmapQueryRunner(QueryRunner):
```

Comment:
This lgtm, but you may want to check later if the ProductAnalytics folks want to make the resulting query a bit more similar to the ones they use, like using `_get_date_subqueries`. That said, I am in no position to check if that is recommended or not; I was just checking the implementation differences between our queries on the `WebAnalyticsQueryRunner` and the `TrendsQueryBuilder` for instance.